### DISCUSSION TITLE
Adopting --incompatible_enable_proto_toolchain_resolution

### DISCUSSION THREAD

**Discussion by alexeagle** [bazelbuild/rules_proto#213] [General]
This flag, introduced in Bazel 7 and advertised in the [6.0.0 release](https://github.com/bazelbuild/rules_proto/releases/tag/6.0.0) of rules_proto, allows users to avoid building the `cc_binary` target `@com_google_protobuf//:protoc` from source. This makes builds faster,  reduces warnings spam in the Bazel log, and in cases where a codebase has no C++ code, it removes the dependency on a functional C++ toolchain.

However, as long as *any* hard-coded references to `@com_google_protobuf//:protoc` are reachable via the dependency graph, the benefit of the flag is negated.

End-users are the ones affected, but are not in a position to fix: they'll need to discover these bad dependency edges, isolate where the problem should be addressed, and find/file an issue on some ruleset. This discussion is meant to collect some practices for doing this, and a list of known issues.

To find why you depend on protoc:

`bazel cquery 'somepath("//...", "@com_google_protobuf//:all")'`

To prevent protoc compilation from source (and grpc while we're at it) add to `.bazelrc`:

```
common --per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --host_per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --host_per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
```

FYI @fmeum and @jiawen who are helping to facilitate the migration.

**Comment by alexeagle** [comment]
KNOWN ISSUE:
 stardoc https://github.com/bazelbuild/stardoc/pull/221

**Comment by fmeum** [comment]
Migration advice: References to `@com_google_protobuf` needed for backwards compatibility with `--noincompatible_enable_proto_toolchain_resolution` can be wrapped in [these helper functions](https://github.com/bazelbuild/rules_proto/blob/d205d37866925569d99b4d6cdcba172326ecf812/proto/proto_common.bzl#L50-L54).

**Comment by alexeagle** [comment]
rules_swift: https://github.com/bazelbuild/rules_swift/issues/1203

**Comment by alexeagle** [comment]
rules_go:
already fixed in https://github.com/bazelbuild/rules_go/pull/3919 - waiting for release

**Comment by alexeagle** [comment]
rules_rust: https://github.com/bazelbuild/rules_rust/issues/2627

**Comment by alexeagle** [comment]
rules_python: was fixed by https://github.com/bazelbuild/rules_python/pull/1577 and released in 0.32.0

**Reply by alexeagle** [reply]
This was reverted by https://github.com/bazelbuild/rules_python/pull/1933 which deleted the code we fixed. Will try again to fix in the protobuf repo :(

**Comment by alexeagle** [comment]
rules_ts: already fixed in https://github.com/aspect-build/rules_ts/pull/608 - Released in v2.4.0

**Comment by iitsdani** [comment]
grpc-java: https://github.com/grpc/grpc-java/issues/11152

**Comment by alexeagle** [comment]
protobuf itself: https://github.com/protocolbuffers/protobuf/issues/17075

**Comment by tsawada** [comment]
I am not sure if this is the right place to report this issue but I found that `proto_library` depends on `@com_google_protobuf//:protoc` in some cases.
More details on https://github.com/protocolbuffers/protobuf/issues/20182
